ARE/components/actuator.irtrans/src/main/java/eu/asterics/component/actuator/irtrans/IrTransInstance.java

```diff
@@ -213,116 +243,96 @@ public void receiveEvent(final String data) {
-        int intPort;
-        InetAddress address;
-        byte[] sendBuf;
-
-        DatagramSocket socketOut = null;
-        DatagramPacket packet;
-
+    private void sendString(String cmdString) {
         try {
-            socketOut = new DatagramSocket();
-
-            stringToSend = stringToSend.trim();
-            int spacePos = stringToSend.indexOf(' ');
-            if (spacePos > 0) {
-                spacePos++;
-                String startStr = stringToSend.substring(0, spacePos);
-                String tailStr = stringToSend.substring(spacePos);
-                tailStr = tailStr.replaceAll("\\s", ""); // remove spaces !
-                stringToSend = startStr + tailStr;
-                sendBuf = stringToSend.getBytes();
-
-                address = InetAddress.getByName(propHostname);
-                intPort = Integer.parseInt(propPort);
-                packet = new DatagramPacket(sendBuf, sendBuf.length, address, intPort);
-                socketOut.send(packet);
-                AstericsErrorHandling.instance.reportInfo(this, "IRTrans sent data: " + stringToSend);
-                socketOut.close();
-
-                // This block just receives the answers, not required
-                /*
-                 * Thread.sleep(100); socketIn = new DatagramSocket(21000);
-                 * byte[] buf = new byte[256]; packet = new DatagramPacket(buf,
-                 * buf.length); socketIn.receive(packet); String received = new
-                 * String(packet.getData(), 0, packet.getLength());
-                 * System.out.println("IRTrans, received: " + received);
-                 * socketIn.close();
-                 */
+            int readTimeout = cmdString.contains("learn") ? READ_TIMEOUT_LEARN_MS : READ_TIMEOUT_SEND_MS;
+            if (socket == null || socket.isClosed() || !socket.isConnected()) {
+                initSocket();
+            }
+            if (socket != null) {
+                sendToSocket(cmdString);
+                String result = readFromSocket(readTimeout);
+                AstericsErrorHandling.instance.reportInfo(this, "IRTrans response: " + result);
+                opOutput.sendData(result.getBytes());
             } else {
-                AstericsErrorHandling.instance.reportError(this,
-                        "The IRTrans - Plugin could not send the command " + stringToSend);
+                opOutput.sendData(ERROR_RESULT.getBytes());
             }
-
         } catch (Exception e) {
+            opOutput.sendData(ERROR_RESULT.getBytes());
+            closeSocket();
             AstericsErrorHandling.instance.reportError(this,
                     "The IRTrans - Plugin could not send data. Please verify that the IRTrans Module is connected and installed, and that the IP-address is correctly specified.");
             AstericsErrorHandling.instance.reportInfo(this, e.toString());
         }
     }
 
-    private void receiveCommand() {
```

Comment:
question: can we really remove this? it is actually a breaking change. before: irtrans module received commands all the time and sent it to output port. now: only responses to a sent command are sent to output port.
@ChrisVeigl what do you think?